kernel/preemption/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+/// so the maximum number of cores it supports is `u8::MAX` (256).
+const MAX_CPU_CORES: usize = u8::MAX as usize;
+
+const ATOMIC_U8_ZERO: AtomicU8 = AtomicU8::new(0);
+
+/// The per-core preemption count, indexed by a CPU core's APIC ID.
+/// 
+/// If a CPU's count is `0`, preemption is enabled.
+/// If a CPU's count is greater than `0`, preemption is disabled.
+static PREEMPTION_COUNT: [AtomicU8; MAX_CPU_CORES] = [ATOMIC_U8_ZERO; MAX_CPU_CORES];
+
+
+/// Prevents preemption (preemptive task switching) from occurring
+/// until the returned guard object is dropped.
+pub fn hold_preemption() -> PreemptionGuard {
+    let apic_id = apic::get_my_apic_id();
+    let prev_val = PREEMPTION_COUNT[apic_id as usize].fetch_add(1, Ordering::Relaxed);
+    // If the previous counter value was 0, that indicates we are transitioning
+    // from preemption being enabled to disabled on this CPU.
+    let preemption_was_enabled = prev_val == 0;
+    // Create a guard here immediately after incrementing the counter,
+    // in order to guarantee that a failure below will drop it and decrement the counter.
+    let guard = PreemptionGuard {
+        apic_id,
+        preemption_was_enabled,
+    };
+
+    if preemption_was_enabled {
+        // When transitioning from preemption being enabled to disabled,
+        // we must disable the local APIC timer used for preemptive task switching.
+        apic::get_my_apic()
+            .expect("BUG: hold_preemption() couldn't get local APIC")
+            .write()
+            .enable_lvt_timer(false);
+    } else if prev_val == u8::MAX {
+        // Overflow occurred and the counter value wrapped around, which is a bug.
+        panic!("BUG: Overflow occurred in the preemption counter for CPU {}", apic_id);
+    }
+
+    guard
+}
+
+
+/// A guard type that ensures preemption is disabled as long as it is held.
+/// 
+/// Call [`hold_preemption()`] to obtain a `PreemptionGuard`.
+/// 
+/// Preemption *may* be re-enabled when this guard is dropped,
+/// but not necessarily so, because other tasks on this CPU 
+/// may also have acquired a `PreemptionGuard`.
+pub struct PreemptionGuard {
+    /// The APIC ID of the CPU on which preemption was held.
+    /// This is not necessary, it's just used for sanity checks or debugging.
+    apic_id: u8,
+    /// Whether preemption was enabled when this guard was created.
+    preemption_was_enabled: bool,
+}
+
+impl PreemptionGuard {
+    /// Returns `true` if preemption was originally enabled
+    /// when this guard was created and disabled them.
+    pub fn preemption_was_enabled(&self) -> bool {
+        self.preemption_was_enabled
+    }
+}
+
+impl Drop for PreemptionGuard {
+    fn drop(&mut self) {
+        let apic_id = apic::get_my_apic_id();
+        assert!(
+            self.apic_id == apic_id,
+            "PreemptionGuard::drop(): BUG: APIC IDs did not match! \
+            This indicates an unexpected task migration across CPUs."
+        );
```

Comment:
To enforce this `PreemptionGuard` should be marked `!Sync`.

Reply:
I now realise I didn't do that in #579.

Reply:
yea that's a good idea. Actually we can go even further and say that it shouldn't be marked `Send` either. Technically that isn't truly necessary since it's not protecting task-specific access to an underlying resource (but a PreemptSafeMutexGuard _should_ be `!Send`), but it's probably a good idea nonetheless.

Reply:
Wait sorry I mixed up send and sync. It can be sync, because the drop handler is only run by the thread owning the guard.

Reply:
Right. done :+1: